tutorials/roofit/rf615_simulation_based_inference.py

```diff
@@ -0,0 +1,231 @@
+## \file
+## \ingroup tutorial_roofit
+## \notebook
+## Use Simulation Based Inference (SBI) in RooFit
+##
+## This tutorial shows how to use SBI in ROOT. As reference distribution we
+## choose a simple uniform distribution. The target distribution is chosen to
+## be gaussian with different mean values.
+## The classifier is trained to discriminate between the reference and target
+## distribution.
+## We see how the neural networks generalize to unknown mean values.
+##
+## We compare the approach of using the likelihood ratio trick to morphing
+##
+## \macro_code
+## \macro_output
+##
+## \date July 2024
+## \author Robin Syring
+
+
+import ROOT
+import numpy as np
+from sklearn.neural_network import MLPClassifier
+
+# The samples used for training the classifier in this tutorial / rescale for more accuracy
+n_samples = 1000
+
+
+# morphing as a baseline 
+def morphing(setting):
+
+
+    # define binning for morphing 
+    bin_mu_x = ROOT.RooBinning(4, 0., 4.)
+    grid = ROOT.RooMomentMorphFuncND.Grid2(bin_mu_x)
+    x_var.setBins(50)
+
+    # number of 'sampled' gaussians, if you change it, adjust the binning properly
+    parampoints = np.arange(5)
+    
+    for i in parampoints:
+        # define the sampled gausians
+        mu_help=  ROOT.RooRealVar(f"mu{i}", f"mu{i}", i)
+        help = ROOT.RooGaussian(f'g{i}', f'g{i}', x_var, mu_help, sigma_var)
+        workspace.Import(help)
+
+        # fill the histograms
+        hist = workspace[f'g{i}'].generateBinned([x_var], n_samples)
+
+        # make sure that every bin is filled and we don't get zero probability
+        for i_bin in range(hist.numEntries()):
+            hist.add(hist.get(i_bin), 1.0)
+
+        # add the pdf to the workspace
+        workspace.Import(ROOT.RooHistPdf(f'histpdf{i}', f'histpdf{i}', [x_var], hist, 1))
+        # pdf = workspace[f'histpdf{i}']
+
+        # add the pdf to the grid
+        grid.addPdf(workspace[f'histpdf{i}'], int(i))
+        
+        # for plotting purposes
+        workspace[f'histpdf{i}'].plotOn(frame1)
+    
+    # uncomment to see input plots
+    # frame1= x_var.frame()
+    # c0= ROOT.TCanvas()
+    # frame1.Draw()
+    # input("")
```

Comment:
```suggestion
    # frame1 = x_var.frame()
    # c0 = ROOT.TCanvas()
    # frame1.Draw()
    # input() # wait for user input to proceed
```
Also the commented-out code should be formatted :)

Reply:
Oh, actually the part where you need to draw on the frame needs to be in the commented-out block too:
```python
    # Uncomment to see input plots
    # frame1 = x_var.frame()
    # for i in parampoints:
    #    workspace[f"histpdf{i}"].plotOn(frame1)
    # c0 = ROOT.TCanvas()
    # frame1.Draw()
    # input() # wait for user input to proceed
```